From 627f478abbef27ec52c67a283ba90f1729bf86ba Mon Sep 17 00:00:00 2001
Message-Id: <627f478abbef27ec52c67a283ba90f1729bf86ba.1651138725.git.thomas@gllm.fr>
In-Reply-To: <dfbbc0d05a7d14b403668da8ce51c95172258a46.1651138725.git.thomas@gllm.fr>
References: <dfbbc0d05a7d14b403668da8ce51c95172258a46.1651138725.git.thomas@gllm.fr>
From: Thomas Guillem <thomas@gllm.fr>
Date: Tue, 10 Aug 2021 11:06:25 +0200
Subject: [PATCH 12/18] audiotrack: rework DynamicsProcessing stop

Specifically disable it before unreferencing the jobject and destroy it
before AudioTrack.

This commit tries to fix a very rare crash with DP.
Message-Id: <b0d5fa154e0fc27f6d368977886348269a527aa8.1628587500.git.thomas@gllm.fr>
---
 modules/audio_output/audiotrack.c | 15 +++++++++------
 1 file changed, 9 insertions(+), 6 deletions(-)

diff --git a/modules/audio_output/audiotrack.c b/modules/audio_output/audiotrack.c
index f271d26772..89783c598b 100644
--- a/modules/audio_output/audiotrack.c
+++ b/modules/audio_output/audiotrack.c
@@ -1439,6 +1439,15 @@ Stop( audio_output_t *p_aout )
     else
         vlc_mutex_unlock(  &p_sys->lock );
 
+    if( p_sys->p_dp )
+    {
+        JNI_CALL_INT( p_sys->p_dp, jfields.DynamicsProcessing.setEnabled, false );
+        CHECK_AT_EXCEPTION( "DynamicsProcessing.setEnabled" );
+
+        (*env)->DeleteGlobalRef( env, p_sys->p_dp );
+        p_sys->p_dp = NULL;
+    }
+
     /* Release the AudioTrack object */
     if( p_sys->p_audiotrack )
     {
@@ -1452,12 +1461,6 @@ Stop( audio_output_t *p_aout )
         p_sys->p_audiotrack = NULL;
     }
 
-    if( p_sys->p_dp )
-    {
-        (*env)->DeleteGlobalRef( env, p_sys->p_dp );
-        p_sys->p_dp = NULL;
-    }
-
     /* Release the timestamp object */
     if( p_sys->timestamp.p_obj )
     {
-- 
2.30.2

